pennylane/templates/subroutines/swap_test.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the ``SWAP Test`` template.
+"""
+import pennylane as qml
+from pennylane.templates.decorator import template
+from pennylane.wires import Wires
+
+
+@template
+def SWAPTest(register1, register2, ancilla):
+
+    r""" A controlled-SWAP subroutine that can be used to create a
+    `SWAP test <https://en.wikipedia.org/wiki/Swap_test>`__ between
+    two registers.
+
+    Args:
+        register1 (Iterable or Wires): The first register of wires passed into the template.
```

Comment:
Why the option for iterable or ints instead of just wires in each case?

Reply:
Wouldn't that mean we are only allowing users to pass something like Wires([0, 1]), rather than [0, 1]?

Reply:
It might be confusing to users that it is named `SWAPTest` and yet it is only a subroutine which "can be used to create a SWAP test" (it confused me)

Reply:
That's definitely true, although the SWAP test is really the only procedure I can see this subroutine being used for (in fact, this template is exactly a SWAP test circuit minus the final measurement of the ancilla, which has to be done by the user).

Would you prefer something like `CSWAPLayer`?

Reply:
I've never actually tried this, but does PL allow you to include measurements in templates?

```python
def SWAPTest(register1, register2, ancilla):
    qml.Hadamard(wires=ancilla)

    for i, reg in enumerate(register1):
        qml.CSWAP(wires=[ancilla, reg, register2[i]])

    qml.Hadamard(wires=ancilla)
    return qml.expval(qml.PauliZ(0))


@qml.qnode(dev)
def circuit(register1=None, register2=None, ancilla=None):
    for wire in register1:
        qml.Hadamard(wires=wire)
    return SWAPTest(register1, register2, ancilla)
```

Reply:
It works 🙀 

```python
In [4]: def test_template(x, wires=None):
   ...:     qml.RX(x, wires=wires[0])
   ...:     qml.CNOT(wires=wires)
   ...:     return qml.expval(qml.PauliZ(wires[0]))
   ...:

In [5]: @qml.qnode(dev)
   ...: def circuit(x):
   ...:     qml.broadcast(qml.Hadamard, pattern="single", wires=[0, 1])
   ...:     return test_template(x, wires=[1, 0])
   ...:

In [6]: circuit(0.5)
Out[6]: 0.0

In [7]: print(circuit.draw())
 0: ──H───────────╭X──┤
 1: ──H──RX(0.5)──╰C──┤ ⟨Z⟩
```

@mariaschuld, this is pretty non-standard though, so interested to get your thoughts

Reply:
Wow, I had no idea you could do this :joy:

Reply:
Maybe we should add the <Z> at the end!

Reply:
```python
def SWAPTest(register1, register2, ancilla):
    qml.Hadamard(wires=ancilla)

    for i, reg in enumerate(register1):
        qml.CSWAP(wires=[ancilla, reg, register2[i]])

    qml.Hadamard(wires=ancilla)
    return qml.expval(qml.PauliZ(ancilla))


@qml.qnode(dev)
def circuit(register1=None, register2=None, ancilla=None):
    for wire in register1:
        qml.Hadamard(wires=wire)
    return SWAPTest(register1, register2, ancilla)
```

Calling this:

```pycon
>>> circuit(register1=[0], register2=[1], ancilla=2)
0.4999999999999999
>>>
>>> print(circuit.draw())
 0: ──H──╭SWAP─────┤
 1: ─────├SWAP─────┤
 2: ──H──╰C─────H──┤ ⟨Z⟩
```

Reply:
> Wow, I had no idea you could do this 😂

Neither did I 🤣

Reply:
I never thought about it, but why not? In some sense this is just a "measurement" template, as every circuit before a measurement could be interpreted as rotating the state into the correct measurement basis. 

I actually love it, did not even cross my mind to provide a swap test template!

Reply:
I really like the idea of thinking about this as a "measurement template" 🙂